nemo_curator/modules/semantic_dedup.py

```diff
@@ -304,63 +328,74 @@ def __call__(self, embeddings_dataset: DocumentDataset):
-        embeddings_df = embeddings_df.reset_index(drop=True)
-        kmeans_centroids_file = os.path.join(
-            self.clustering_output_dir, "kmeans_centroids.npy"
-        )
-        np.save(kmeans_centroids_file, centroids)
-        self.logger.info("Saving centroids complete")
-        del kmeans, cupy_darr, centroids
+            self.logger.info(
+                "Computing nearest centroids + distance to centers using kmeans.predict"
+            )
+            t0 = time.time()
+            nearest_cents = kmeans.predict(cupy_darr)
+            self.logger.info(f"Time taken for KMeans Predict: {time.time() - t0}")
+
+            embeddings_df["nearest_cent"] = nearest_cents.astype(np.int32)
+            del nearest_cents
+            meta_df = embeddings_df._meta.copy()
+            meta_df["dist_to_cent"] = cp.zeros(1)
+            embeddings_df = embeddings_df.map_partitions(
+                add_dist_to_cents,
+                embedding_col=self.embedding_col,
+                centroids=kmeans.cluster_centers_,
+                meta=meta_df,
+            )
+            embeddings_df = embeddings_df.reset_index(drop=True)
+            centroids = kmeans.cluster_centers_
+            kmeans_centroids_file = os.path.join(
+                self.clustering_output_dir, "kmeans_centroids.npy"
+            )
+            np.save(kmeans_centroids_file, centroids)
+            self.logger.info("Saving centroids complete")
+            del kmeans, cupy_darr, centroids
 
-        clustering_output_dir = os.path.join(
-            self.clustering_output_dir, "embs_by_nearest_center"
-        )
-        if os.path.exists(clustering_output_dir):
-            self.logger.warning(
-                f"Output directory {clustering_output_dir} already exists and will be overwritten"
+            clustering_output_dir = os.path.join(
+                self.clustering_output_dir, "embs_by_nearest_center"
             )
-            shutil.rmtree(clustering_output_dir)
+            if os.path.exists(clustering_output_dir):
+                self.logger.warning(
+                    f"Output directory {clustering_output_dir} already exists and will be overwritten"
+                )
+                shutil.rmtree(clustering_output_dir)
 
-        embeddings_df.to_parquet(
-            clustering_output_dir,
-            index=False,
-            partition_on="nearest_cent",
-        )
-        self.logger.info(
-            f"Saved embeddings by nearest center to {clustering_output_dir}"
-        )
-        del embeddings_df
+            embeddings_df.to_parquet(
+                clustering_output_dir,
+                index=False,
+                partition_on="nearest_cent",
+            )
+            # TODO question for reviewer do we need a logger time here too? if so what "step" would that be for?
+            self.logger.info(
+                f"Saved embeddings by nearest center to {clustering_output_dir}"
+            )
```

Comment:
What do you mean by logger time ?

Reply:
I mean, in most places where we're logging the time taken at each "micro-stage" within a bigger pipeline so that we can closely observe the slower / faster parts. Given there is so much happening in `ClusteringModel.__call__`, similar to the profiling question above, which parts should we time? I've already added logs for `KMeans.fit` and `KMeans.predict`.

For now I'm calling the step `Assigning distance to each embedding` but pls let me know if that's not accurate.